flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/operations/RichTableSourceQueryOperation.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.operations;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.plan.stats.FlinkStatistic;
+import org.apache.flink.table.sources.TableSource;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link TableSourceQueryOperation} with {@link FlinkStatistic} and qualifiedName.
+ */
+@Internal
+public class RichTableSourceQueryOperation<T> extends TableSourceQueryOperation<T> {
+	private final FlinkStatistic statistic;
+	private List<String> qualifiedName;
```

Comment:
Why do we need the qualifiedName? Where does the qualified name comes from? The purpose of having TableSourceQueryOperation/DataStreamQueryOperation to not have a connection to the catalog, as they are created when e.g. `fromTableSource(TableSource)` or `fromDataStream(DataStream)` which is an inline operation.

If I understand it correctly this is needed for your tests to have a deterministic naming. To be honest I think this shows a problem in the tests infrastructure. As a workaround though I would change the `List<String> qualifiedName` to `String uniqueId`.

Reply:
the `qualifiedName` is used for creating `FlinkRelOptTable` in `QueryOperationConverter` class, and the plan could show given table name while registering table source.
the qualified name is from `DatabaseCalciteSchema`:
`
			if (table instanceof QueryOperationCatalogView) {
				QueryOperationCatalogView view = (QueryOperationCatalogView) table;
				QueryOperation operation = view.getQueryOperation();
				if (operation instanceof DataStreamQueryOperation) {
					List<String> qualifiedName = Arrays.asList(catalogName, databaseName, tableName);
					((DataStreamQueryOperation) operation).setQualifiedName(qualifiedName);
				} else if (operation instanceof RichTableSourceQueryOperation) {
					List<String> qualifiedName = Arrays.asList(catalogName, databaseName, tableName);
					((RichTableSourceQueryOperation) operation).setQualifiedName(qualifiedName);
				}
				return QueryOperationCatalogViewTable.createCalciteTable(view);
			} else if (table instanceof ConnectorCatalogTable) {
...
`
for unregister table source (`fromTableSource(TableSource)`) or data stream (`fromDataStream(DataStream)`), their name is "unregistered_XXX" which is same with flink planner.

Reply:
What do you mean it is lost? The name there is a name of a view (`QueryOperationCatalogView` is a view). Views are expanded, so I think it totally makes sense that it is not present in the explain.

I would say the blink behavior is misleading as you are overwriting original name with a name of a view.

Reply:
Ok, but I think I get your point as it changes the old behavior slightly when registered via: `registerDataStream(String name, DataStream)`. Let me think about it for a second.

Reply:
To be honest. I see pros & cons of that solution. 

My personal opinion is that `DataStreamQueryOperation` and `TableSourceQueryOperation` are inline operations and are always registered as views, thus expanded during optimization if accessed via catalog.

How much of an effort would it be to have a consistent behavior with flink planner for this PR? I would prefer to move the discussion what name/digest to use to a separate issue, if possible.

Reply:
about 7000+ test cases will be changed if blink-planner has a consistent behavior with flink planner.  My suggestion is: I add a `FIXME` about `qualifiedName`, and fix them in another issue.

Reply:
I thought so :( I think this is a reasonable trade off for now.